### PR TITLE
Ready: Comm share fixes

### PR DESCRIPTION
**Pull request by ArchShaman**
Cleaned version of the previous PR after disasterous accidental copy paste of some widget code.

### CODE REVIEW

File: LuaRules/Gadgets/game_share_mode.lua [hunk 1]
```diff
@@ -236,33 +238,52 @@ if (gadgetHandler:IsSyncedCode()) then
 		end
 	end
 	
+	function gadget:GameStart()
+		Spring.SetGameRulesParam("sharemode",config.mergetype)
```
Comment (sprunk) [inline LuaRules/Gadgets/game_share_mode.lua]: Redundant. Already accessible from modoptions.

File: LuaRules/Gadgets/game_share_mode.lua [hunk 2]
```diff
@@ -236,33 +238,52 @@ if (gadgetHandler:IsSyncedCode()) then
...
+						invites[key] = nil
+					end
+					if data and invitestring ~= "" and data["timeleft"] > 0 then
+						invitestring = invitestring .. ", " .. data["id"] .. " " .. data["timeleft"] .. " " .. data["controller"]
+					else
+						invitestring = data["id"] .. " " .. data["timeleft"] .. " " .. data["controller"]
 					end
 				end
+				--Spring.Echo("DEBUG: Got Invitestring: " .. invitestring)
+				Spring.SetTeamRulesParam(GetTeamID(player),"invites",invitestring,{private=true})
```
Comment (sprunk) [inline LuaRules/Gadgets/game_share_mode.lua]: Split this up into objects and then into fields. For example:

```
commshare_invite_count
commshare_invite_1_id
commshare_invite_1_timeleft
```

Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: Alright. I'll do it. Would something like this help?
commshare_invitecount = 2
commshare_invite1 = "id=2, timeleft=45, controller=1"

File: LuaRules/Gadgets/game_share_mode.lua (hunk 1 above)
Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: I had trouble reading the modoptions on my widget. It would report that commshare = nil

**Comment by ArchShaman** [comment]
Any other suggestions? Or looks okay? 

I did testing in sterile environment. All the commands should be working okay. Will do some live tests to see if everything's running okay today.

File: LuaRules/Gadgets/game_share_mode.lua [hunk 3]
```diff
@@ -138,19 +122,21 @@ if (gadgetHandler:IsSyncedCode()) then
...
 		end
 		local originalteam = GetTeamID(playerid)
 		local name,_,spec  = Spring.GetPlayerInfo(playerid)
 		if Spring.AreTeamsAllied(originalteam,target) and spec == false and target ~= Spring.GetGaiaTeamID() and config.mergetype ~= "none" then
-			Spring.Echo("Commshare: Assigning player id " .. playerid .. "(" .. name .. ") to team " .. target)
+			Spring.Echo("[Commshare] Assigning player id " .. playerid .. "(" .. name .. ") to team " .. target)
 			local name,_,spec,_,_,allyteam = Spring.GetPlayerInfo(playerid)
 			if GetSquadSize(originalteam) - 1 == 0 then
-				MergeUnits(originalteam,target)
+				local metal = select(1,Spring.GetTeamResources(originalteam,"metal"))
```
Comment (sprunk) [inline LuaRules/Gadgets/game_share_mode.lua]: select(1,x) is just x

File: LuaRules/Gadgets/game_share_mode.lua [hunk 4]
```diff
@@ -238,31 +223,47 @@ if (gadgetHandler:IsSyncedCode()) then
...
+					invitecount = invitecount+1
+					if data["timeleft"] > 0 then
+						data["timeleft"] = data["timeleft"] - 1
+					end
+					if data["timeleft"] == -1 then -- this is so we know an invite has expired. UI will remove the invite at timeleft = 0, but gadget will remove it at -1. Otherwise UI will just see it constantly at 1.
+						invitecount = invitecount-1
+						invites[key] = nil
+					end
+					if data and data["timeleft"] > -1 then
+						Spring.SetTeamRulesParam(GetTeamID(player),"commshare_invite_"..invitecount.."_timeleft",data["timeleft"],{private=true})
```
Comment (sprunk) [inline LuaRules/Gadgets/game_share_mode.lua]: Make a single global instance of the {private = true} table and pass that (currently it creates a new table each invocation).

File: LuaRules/Gadgets/game_share_mode.lua [hunk 5]
```diff
@@ -181,48 +167,47 @@ if (gadgetHandler:IsSyncedCode()) then
...
 			if Invites[player][target]["controller"] ~= player then
-				MergePlayer(target,GetTeamID(player))
+				MergePlayer(player,GetTeamID(target))
 			else -- target->player
 				MergeTeams(GetTeamID(target),GetTeamID(player))
 				teamlist = nil
 			end
 			Invites[player][target] = nil
-			Invites[target][player] = nil
+			if Invites[target] and Invites[target][player] then Invites[target][player] = nil end
```
Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: Ah i see. Thanks.

File: LuaRules/Gadgets/game_share_mode.lua (hunk 4 above)
Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: Eww... really? Yuck.

**Comment by ArchShaman** [comment]
This _should_ be stable. Internal testing shows all commands work fine, nothing seems to error out. Marking this as ready, baring any further desired changes.

File: LuaRules/Gadgets/game_share_mode.lua (hunk 3 above)
Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: lol.. That's a stupid error of mine.

File: LuaRules/Gadgets/game_share_mode.lua [hunk 6]
```diff
@@ -138,19 +124,21 @@ if (gadgetHandler:IsSyncedCode()) then
...
 		end
 		local originalteam = GetTeamID(playerid)
 		local name,_,spec  = Spring.GetPlayerInfo(playerid)
 		if Spring.AreTeamsAllied(originalteam,target) and spec == false and target ~= Spring.GetGaiaTeamID() and config.mergetype ~= "none" then
-			Spring.Echo("Commshare: Assigning player id " .. playerid .. "(" .. name .. ") to team " .. target)
+			Spring.Echo("[Commshare] Assigning player id " .. playerid .. "(" .. name .. ") to team " .. target)
 			local name,_,spec,_,_,allyteam = Spring.GetPlayerInfo(playerid)
 			if GetSquadSize(originalteam) - 1 == 0 then
-				MergeUnits(originalteam,target)
+				local metal,_ = Spring.GetTeamResources(originalteam,"metal")
```
Comment (sprunk) [inline LuaRules/Gadgets/game_share_mode.lua]: No need to assign fake values after the last real one either:
`local metal = Spring.GetTeamResources(originalteam,"metal")`

**Comment by GoogleFrog** [comment]
@sprunk you have more recent reviews. Merge if you would like to.

**Comment by GoogleFrog** [comment]
I'll just mess with it if it doesn't work.

**Pull request merged**